src/renderers/webgl/WebGLBindingStates.js

```diff
@@ -0,0 +1,393 @@
+﻿/**
+ * @author Mugen87 / https://github.com/Mugen87
+ * @author Takahiro / https://github.com/takahirox
+ */
+
+function WebGLBindingStates( gl, extensions, state, attributes, capabilities ) {
+
+	var maxVertexAttributes = gl.getParameter( gl.MAX_VERTEX_ATTRIBS );
+
+	var extension = extensions.get( 'OES_vertex_array_object' );
+	var vaoAvailable = capabilities.isWebGL2 || extension !== null;
+
+	var bindingStates = new Map();
+
+	var defaultState = createBindingState( null );
+	var currentState = defaultState;
+
+	function setup( material, program, geometry, hasMorphTarget ) {
+
+		var state = getBindingState( geometry, program, material );
+
+		if ( currentState !== state ) {
+
+			currentState = state;
+			bindVertexArrayObject( currentState.object );
+
+		}
+
+		var updateBuffers = hasMorphTarget;
+
+		if ( vaoAvailable ) {
+
+			if ( geometry.version > currentState.version ) {
+
+				currentState.version = geometry.version;
+
+				updateBuffers = true;
+
+			}
+
+		} else {
+
+			var wireframe = material.wireframe === true;
+
+			if ( currentState.geometry !== geometry.id ||
+				currentState.program !== program.id ||
+				currentState.wireframe !== wireframe ) {
+
+				currentState.geometry = geometry.id;
+				currentState.program = program.id;
+				currentState.wireframe = wireframe;
+
+				updateBuffers = true;
+
+			}
+
+		}
+
+		if ( updateBuffers ) {
```

Comment:
Following the discussion https://github.com/mrdoob/three.js/pull/14590#pullrequestreview-141652383

> What happens if the user change a attribute data? We have to update the VAO. It would be a bad usage of the library maybe.

Regarding `BufferGeometry`'s `stride/offset/count`, I don't wanna fine-grain update check for the performance and simple code. So I'm thinking of a workaround that is asking users to explicitly call `geometry.version ++;` if users update them on the fly.

@DavidPeicho @Mugen87

Reply:
An other more strict option is to think of these properties as immutable. If the user wants to change them, it's necessary to create a new attribute and add it via `BufferGeometry.addAttribute()`. In this case, `geometry.version` would be updated automatically. 

In any event, the library would lose flexibility. Especially since we assume `.addAttribute()` is actually used. Right now, you can change `BufferGeometry.attributes` directly without breaking anything.

Reply:
In the meanwhile, I think it's okay to accept such a change. I personally think the performance gain due to VAO usage is more important than the mentioned flexibility with buffer attributes.